### DISCUSSION TITLE
There are multiple different ways to configure file change events for language clients in an extension.  Which one is "correct"?

### DISCUSSION THREAD

**Discussion by thyeggman** [microsoft/vscode-discussions#511] [Extension Development QnA] [answered]
I am currently working on a language server that integrates with an extension.  There appear to be two different ways to handle file change events.  The first one, and the one that we currently implement in our language server, looks like this:

```ts
documents.onDidChangeContent(change => {
  // Do validation...
});
```

In our extension, we pass the following `LanguageClientOption`:

```ts
  const clientOptions: LanguageClientOptions = {
    documentSelector: [{ pattern: "pattern..." }],
    // ...
  };
```

However, when implementing a feature, I ran across a different setup that appears to behave identically.  We can set up this event in our language server:

```ts
  connection.onDidChangeWatchedFiles(change => {
    change.changes.forEach(change => {
      if (change.type === FileChangeType.Changed) {
        // Do validation...
      }
    });
  });
```

In our extension, we would then add this `LangaugeClientOption`:

```ts
  const clientOptions: LanguageClientOptions = {
    synchronize: {
      fileEvents: vscode.workspace.createFileSystemWatcher("pattern...")
    }
  };
```

The [example](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide) in the language server documentation uses `onDidChangeContent` and puts its validation there, and that's what we currently do.  The example also configures `onDidChangeWatchedFiles`, but just logs at that point.

Is there a difference to using the second method?  Should they be used for different purposes?  Is one preferred for doing validation over another?

**Comment by osortega** [comment]
Not 100% sure about the differences on both listeners but seems that the preferred method by VS Code might be `onDidChangeWatchedFiles`? 

@dbaeumer might be the right person to ask

**Comment by dbaeumer** [comment, marked as answer]
The second version is to watch for files changes that are manipulated outside of VS Code (e.g. using vim for example). If the file is not open in VS Code itself the server will not receive a content change event from VS Code in these cases.

**Comment by nachiket87** [comment]
@thyeggman Hey! I'm experiencing a similar issue. How did you end up handling this in your LSP?
